#### DS 862 Lecture 4 Practice Code

In [1]:
# usual imports
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
%matplotlib notebook
from sklearn.model_selection import train_test_split
# Each is a different implemntation of a text transform tool: Bag of Words & Tfidf
from sklearn.feature_extraction.text import CountVectorizer,TfidfVectorizer


#### Let's first play with Yelp data. Earlier, we performed sentiment analysis on this dataset using Random Forest.  

In [2]:
Yelp_data = pd.read_csv('yelp_labelled.txt' , sep = "\t", names = ['text', 'sentiment'])
Yelp_data.dropna(inplace = True)
Yelp_data.head()

text  sentiment
0                            Wow... Loved this place.        1.0
3                                  Crust is not good.        0.0
4           Not tasty and the texture was just nasty.        0.0
10  Stopped by during the late May bank holiday of...        1.0
11  The selection on the menu was great and so wer...        1.0

#### Split data to 80% training and 20% test set. (Use Random State  = 24) 

In [3]:
X = Yelp_data.drop('sentiment',axis=1)
y = Yelp_data['sentiment']

In [4]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=24)

#### Here are few libararies we do need from here on

In [5]:
from sklearn.pipeline import Pipeline
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import GridSearchCV
from sklearn.naive_bayes import BernoulliNB

#### Use Pipeline and define CountVectorizer() as 'vect' and MultiNomial Naive Bayes as your 'clf' - classifier. Then set your parameters to

'vect__min_df':[1, 2, 3, 5, 10], 

'vect__max_df':[50, 100, 150, 200, 500, 1000, 1200], 

'clf__alpha':[0, 0.1, 0.2, 0.5, .8, 1]


In [6]:
nb_clf = Pipeline([('vect', CountVectorizer()),
                   ('clf', MultinomialNB())])

parameters = {'vect__min_df':[1, 2, 3, 5, 10],
              'vect__max_df':[50, 100, 150, 200, 500, 1000, 1200],
              'clf__alpha':[0, 0.1, 0.2, 0.5, .8, 1]}

#### Using GridSearchCV find the best parameters and use it to calculate test error. Did you beat Random Forest?

##### Note to self: code did not compile when X_train was a pd df, changing it to a series worked; that code compiled a year ago without slicing the data frame, do some reading about data types, particularly why pd dfs isn't working with sklearn

In [17]:
gs_clf = GridSearchCV(nb_clf, parameters, cv = 10, n_jobs = -1)
gs_fit = gs_clf.fit(X_train['text'],y_train)

In [19]:
gs_fit.score(X_test['text'],y_test)

0.755

In [20]:
gs_fit.best_params_

{'clf__alpha': 1, 'vect__max_df': 100, 'vect__min_df': 1}

#### Use Pipeline and define CountVectorizer() as 'vect' and Bernoulli Naive Bayes as your 'clf' - classifier. Then set your parameters to

'vect__min_df':[1, 2, 3, 5, 10], 

'vect__max_df':[50, 100, 150, 200, 500, 1000, 1200], 

'clf__alpha':[0, 0.1, 0.2, 0.5, .8, 1]

In [24]:
nb_clf = Pipeline([('vect', CountVectorizer()),
                   ('clf', BernoulliNB())])

parameters = {'vect__min_df':[1, 2, 3, 5, 10],
              'vect__max_df':[50, 100, 150, 200, 500, 1000, 1200],
              'clf__alpha':[0, 0.1, 0.2, 0.5, .8, 1]}

In [27]:
gs_clf = GridSearchCV(nb_clf, parameters, cv = 10, n_jobs = -1)
gs_fit = gs_clf.fit(X_train['text'],y_train)

#### Using GridSearchCV find the best parameters and use it to calculate test error. Did you beat Random Forest?


In [28]:
gs_fit.best_params_

{'clf__alpha': 0.2, 'vect__max_df': 200, 'vect__min_df': 2}

In [29]:
gs_fit.score(X_test['text'],y_test)

0.73

#### What parameters are chosen by GridSearchCV?

alpha of 0.2, max document frequency of 200, min document frequency of 2

#### Now it's time for a new dataset! Let's play with SMS dataset. We would like to develop a model by which filter spam/ham text messages. Let's explore this dataset first.

In [30]:
import pandas as pd
col_names = ['label', 'message']
smsData = pd.read_csv('SMSSpamCollection.tsv', sep='\t', header = 0, names = col_names)
smsData.head(5)

label                                            message
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [31]:
smsData.shape

(5572, 2)

#### Repeat the procedure you applied on Yelp data on SMS data. Can you get better results by using Bernoulli Naive Bayes or MultiNomial Naive Bayes? What is the best score on test set using best tuning parameters? 

In [33]:
sms_data = pd.read_csv('SMSSpamCollection.tsv' , sep = "\t", names = ['spam?', 'text'])
sms_data.dropna(inplace = True)
sms_data = sms_data[1:]
sms_data.head()

spam?                                               text
1   ham  Go until jurong point, crazy.. Available only ...
2   ham                      Ok lar... Joking wif u oni...
3  spam  Free entry in 2 a wkly comp to win FA Cup fina...
4   ham  U dun say so early hor... U c already then say...
5   ham  Nah I don't think he goes to usf, he lives aro...

In [39]:
X = sms_data['text']
y = sms_data['spam?']

In [40]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=24)

In [41]:
nb_clf = Pipeline([('vect', CountVectorizer()),
                   ('clf', MultinomialNB())])

parameters = {'vect__min_df':[1, 2, 3, 5, 10],
              'vect__max_df':[50, 100, 150, 200, 500, 1000, 1200],
              'clf__alpha':[0, 0.1, 0.2, 0.5, .8, 1]}

In [42]:
gs_clf = GridSearchCV(nb_clf, parameters, cv = 10, n_jobs = -1)
gs_fit = gs_clf.fit(X_train,y_train)

In [43]:
gs_fit.score(X_test,y_test)

0.9847533632286996

In [44]:
gs_fit.best_params_

{'clf__alpha': 0.8, 'vect__max_df': 500, 'vect__min_df': 1}

In [45]:
nb_clf = Pipeline([('vect', CountVectorizer()),
                   ('clf', BernoulliNB())])

parameters = {'vect__min_df':[1, 2, 3, 5, 10],
              'vect__max_df':[50, 100, 150, 200, 500, 1000, 1200],
              'clf__alpha':[0, 0.1, 0.2, 0.5, .8, 1]}

In [46]:
gs_clf = GridSearchCV(nb_clf, parameters, cv = 10, n_jobs = -1)
gs_fit = gs_clf.fit(X_train,y_train)

In [47]:
gs_fit.score(X_test,y_test)

0.9847533632286996

In [48]:
gs_fit.best_params_

{'clf__alpha': 0.1, 'vect__max_df': 500, 'vect__min_df': 2}

For the values of the tuning parameters used in the grid search cross validation, Bernoulli and Multinomial Naive Bayes produce the same validation set accuracy. The models use a different combination of tuning parameters, but both produce a test set accuracy of ~98.48%.